File: android/app/src/main/java/com/betterlise/app/data/api/ApiError.kt
package com.betterlise.app.data.api

sealed class ApiError(message: String) : Exception(message) {
    data object Unauthorized : ApiError("Ta session a expiré. Reconnecte-toi.")
    data object SessionExpired : ApiError("Ta session a expiré. Reconnecte-toi.")
    data class InvalidCredentials(val detail: String) : ApiError(detail)
    data class Validation(val detail: String) : ApiError(detail)
    data class RateLimited(val detail: String) : ApiError(detail)
    data class Server(val code: String, val detail: String) : ApiError(
        if (code == "LISE_UNAVAILABLE") {
            "Lise ne répond pas pour le moment. Réessaie dans quelques instants."
        } else {
            "Une erreur est survenue sur le serveur."
        },
    )
    data class Network(val detail: String?) : ApiError("Impossible de joindre Better Lise. Vérifie ta connexion.")
    data object Decoding : ApiError("Réponse inattendue du serveur.")

    /** Errors that a fresh login can fix. */
    val requiresReauthentication: Boolean get() = this is Unauthorized || this is SessionExpired

    companion object {
        fun from(body: ApiErrorBody): ApiError = when (body.code) {
            "UNAUTHORIZED" -> Unauthorized
            "SESSION_EXPIRED" -> SessionExpired
            "INVALID_CREDENTIALS" -> InvalidCredentials(body.message)
            "VALIDATION" -> Validation(body.message)
            "RATE_LIMITED" -> RateLimited(body.message)
            else -> Server(body.code, body.message)
        }
    }
}

File: ios/BetterLise/Core/API/APIError.swift
import Foundation

enum APIError: Error, Equatable, LocalizedError {
    case unauthorized
    case sessionExpired
    case invalidCredentials(String)
    case validation(String)
    case rateLimited(String)
    case server(code: String, message: String)
    case network(String)
    case decoding

    init(body: APIErrorBody) {
        switch body.code {
        case "UNAUTHORIZED": self = .unauthorized
        case "SESSION_EXPIRED": self = .sessionExpired
        case "INVALID_CREDENTIALS": self = .invalidCredentials(body.message)
        case "VALIDATION": self = .validation(body.message)
        case "RATE_LIMITED": self = .rateLimited(body.message)
        default: self = .server(code: body.code, message: body.message)
        }
    }

    /// Errors that a fresh login can fix.
    var requiresReauthentication: Bool {
        self == .unauthorized || self == .sessionExpired
    }

    var errorDescription: String? {
        switch self {
        case .unauthorized, .sessionExpired:
            "Ta session a expiré. Reconnecte-toi."
        case .invalidCredentials(let message), .validation(let message), .rateLimited(let message):
            message
        case .server(let code, _) where code == "LISE_UNAVAILABLE":
            "Lise ne répond pas pour le moment. Réessaie dans quelques instants."
        case .server:
            "Une erreur est survenue sur le serveur."
        case .network:
            "Impossible de joindre Better Lise. Vérifie ta connexion."
        case .decoding:
            "Réponse inattendue du serveur."
        }
    }
}
